Finish removing the BigInts from * for FD{Int128}!

NHDaly opened the pull request:
Finally implements the fast-multiplication optimization from https://github.com/JuliaMath/FixedPointDecimals.jl/pull/45! :)

This PR optimizes multiplication for `FixedDecimal{Int64}` and `FixedDecimal{Int128}`. In the process, we also undid an earlier optimization which is no longer needed after julia 1.8, and that makes multiplication about 2x fast for the smaller int types as well! 🎉 

This is a follow-up to
https://github.com/JuliaMath/FixedPointDecimals.jl/pull/93, which introduces an Int256 type for widemul. However after that PR, the fldmod still required 2 BigInt allocations.

Now, this PR uses a custom implementation of the LLVM div-by-const optimization for (U)Int128 and for (U)Int256, which briefly widens to Int512 (😅), to perform the fldmod by the constant 10^f coefficient.

After this PR, FD multiply performance scales linear with the number of bits. FD{Int128} has no allocations, and is only 2x slower than 64-bit. :) And it makes all other multiplications ~2x faster.

master:
```julia
julia> using FixedPointDecimals, BenchmarkTools

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int16,3}(1.234))
  83.750 μs (0 allocations: 0 bytes)
FixedDecimal{Int16,3}(0.000)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int32,3}(1.234))
  84.916 μs (0 allocations: 0 bytes)
FixedDecimal{Int32,3}(1700943.280)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int64,3}(1.234))
  249.083 μs (0 allocations: 0 bytes)
FixedDecimal{Int64,3}(4230510070790917.029)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int128,3}(1.234))
  4.660 ms (248829 allocations: 4.70 MiB)
FixedDecimal{Int128,3}(-66726338547984585007169386718143307.324)

# unsigned

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt16,3}(1.234))
  56.791 μs (0 allocations: 0 bytes)
FixedDecimal{UInt16,3}(0.000)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt32,3}(1.234))
  60.000 μs (0 allocations: 0 bytes)
FixedDecimal{UInt32,3}(4191932.283)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt64,3}(1.234))
  172.958 μs (0 allocations: 0 bytes)
FixedDecimal{UInt64,3}(16576189118051436.703)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt128,3}(1.234))
  5.408 ms (308621 allocations: 6.14 MiB)
FixedDecimal{UInt128,3}(303384805088638153637410092093845905.434)
```
After PR #93:
```julia
julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int16,3}(1.234))
  83.750 μs (0 allocations: 0 bytes)
FixedDecimal{Int16,3}(0.000)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int32,3}(1.234))
  85.000 μs (0 allocations: 0 bytes)
FixedDecimal{Int32,3}(1700943.280)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int64,3}(1.234))
  248.958 μs (0 allocations: 0 bytes)
FixedDecimal{Int64,3}(4230510070790917.029)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int128,3}(1.234))
  4.673 ms (160798 allocations: 3.22 MiB)
FixedDecimal{Int128,3}(-66726338547984585007169386718143307.324)

# unsigned

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt16,3}(1.234))
  56.791 μs (0 allocations: 0 bytes)
FixedDecimal{UInt16,3}(0.000)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt32,3}(1.234))
  60.041 μs (0 allocations: 0 bytes)
FixedDecimal{UInt32,3}(4191932.283)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt64,3}(1.234))
  173.000 μs (0 allocations: 0 bytes)
FixedDecimal{UInt64,3}(16576189118051436.703)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt128,3}(1.234))
  4.750 ms (190708 allocations: 4.82 MiB)
FixedDecimal{UInt128,3}(303384805088638153637410092093845905.434)

```
After this PR:
```julia
julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int16,3}(1.234))
  48.458 μs (0 allocations: 0 bytes)
FixedDecimal{Int16,3}(0.000)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int32,3}(1.234))
  57.000 μs (0 allocations: 0 bytes)
FixedDecimal{Int32,3}(1700943.280)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int64,3}(1.234))
  90.708 μs (0 allocations: 0 bytes)
FixedDecimal{Int64,3}(4230510070790917.029)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{Int128,3}(1.234))
  180.125 μs (0 allocations: 0 bytes)
FixedDecimal{Int128,3}(-66726338547984585007169386718143307.324)

# unsigned

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt16,3}(1.234))
  42.708 μs (0 allocations: 0 bytes)
FixedDecimal{UInt16,3}(0.000)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt32,3}(1.234))
  51.250 μs (0 allocations: 0 bytes)
FixedDecimal{UInt32,3}(4191932.283)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt64,3}(1.234))
  80.042 μs (0 allocations: 0 bytes)
FixedDecimal{UInt64,3}(16576189118051436.703)

julia> @btime for _ in 1:10000 fd = fd * fd end setup = (fd = FixedDecimal{UInt128,3}(1.234))
  162.417 μs (0 allocations: 0 bytes)
FixedDecimal{UInt128,3}(303384805088638153637410092093845905.434)
```

NHDaly committed:
Use Int256 to avoid BigInt in FD operations.

We do not here explicitly introduce support for FD{BitIntegers.Int256},
though that should work out of the box both before and after this PR.

Rather, this PR _uses_ a (U)Int256 under the hood to prevent allocations
from Int128 widening to BigInt in FD operations.

NHDaly committed:
Further reduce BigInts by skipping a `rem()` in iseven

NHDaly committed:
Fix ambiguity in _widemul(Int256, UInt256)

NHDaly committed:
Bump patch version number

NHDaly committed:
Add compat for BitIntegers

NHDaly committed:
Finish removing the BigInts from * for FD{Int128}!

Finally implements the fast-multiplication optimization from
https://github.com/JuliaMath/FixedPointDecimals.jl/pull/45, but this
time for 128-bit FixedDecimals! :)

This is a follow-up to
https://github.com/JuliaMath/FixedPointDecimals.jl/pull/93, which
introduces an Int256 type for widemul. However, the fldmod still
required 2 BigInt allocations.

Now, this PR uses a custom implementation of the LLVM div-by-const
optimization for (U)Int256, which briefly widens to Int512 (😅) to
perform the fldmod by the constant 10^f coefficient.

This brings 128-bit FD multiply to the same performance as 64-bit. :)

NHDaly commented:
Huzzah! I finally feel like we can bring the ideas from https://github.com/JuliaMath/FixedPointDecimals.jl/pull/45 to this package, and it's still valuable. It turns out that LLVM has already applied this optimization automatically for Int128 div by const (thus * for `FixedDecimal{Int64}` is already fast). But this PR adds support for `FixedDecimal{Int128}`, which needed a custom function for Int256 div by const. :)

See here, for my note on FixedDecimal{Int64} support:
https://github.com/JuliaMath/FixedPointDecimals.jl/pull/45#issuecomment-2163836966

NHDaly committed:
Support older versions of julia

NHDaly commented:
Now that there's julia's effects system, this doesn't need `@pure`, so it's actually possible to do this! 
And now that we've introduced BitIntegers.jl, the code is way simpler. 😊 woohoo!

NHDaly committed:
Comments

NHDaly committed:
Disable fldmod-by-const tests on older julia

NHDaly committed:
Fix one other case of iseven allocating a BigInt

NHDaly committed:
Apply this optimization to FD{Int64} as well.

omus commented on test/runtests.jl:
```diff
@@ -10,4 +10,12 @@ include(joinpath(pkg_path, "test", "utils.jl"))
 @testset "FixedPointDecimals" begin
     include("FixedDecimal.jl")
-end  # global testset
+end
+
+
+if isdefined(Base, Symbol("@assume_effects"))
+@testset "FixedPointDecimals" begin
+    include("fldmod-by-const_tests.jl")
+end
+end
```
Why not include this in the existing FixedPointDecimals testset?
```suggestion
    isdefined(Base, Symbol("@assume_effects")) && include("fldmod-by-const_tests.jl")
end
```

NHDaly replied:
Oh, oops, I meant to rename the test-set to `fldmod-by-const_tests`.

🤷 i dunno, i guess it seems nice to have a separate test-set per file, so that you can group the failures? but you're right that we probably _also_ want an outer testset so that a failure in one doesn't abort the rest of the tests.. annoying.

I'll take your change, i think that's preferable.

omus commented on src/FixedPointDecimals.jl:
```diff
@@ -36,9 +36,22 @@ export checked_abs, checked_add, checked_cld, checked_div, checked_fld,
 using Base: decompose, BitInteger
 
-import BitIntegers  # For 128-bit _widemul / _widen
+using BitIntegers: BitIntegers, UInt256, Int256
```
Why not import the other used types?
```suggestion
using BitIntegers: BitIntegers, Int256, Int512, UInt256, UInt512
```

NHDaly replied:
Seems reasonable. The (U)Int256 were used much more heavily than (U)Int512. So I was inclined to refer to them directly through the qualified name. But I think your suggestion is better.

omus commented on src/fldmod-by-const.jl:
```diff
@@ -0,0 +1,98 @@
+# Annoyingly, Unsigned(T) isn't defined for BitIntegers types:
+# https://github.com/rfourquet/BitIntegers.jl/pull/2
+_unsigned(x) = unsigned(x)
+_unsigned(::Type{Int256}) = UInt256
```
Is there a missing test?
```suggestion
_unsigned(::Type{Int256}) = UInt256
_unsigned(::Type{Int512}) = UInt512
```

NHDaly replied:
No, the built-in `unsigned(T)` already worked for the UInt256, and I think this isn't needed for UInt512, since we are only widening to 512 _after_ we call unsigned.

```julia
julia> unsigned(FixedPointDecimals.UInt256)
BitIntegers.UInt256

julia> unsigned(FixedPointDecimals.Int256)
ERROR: MethodError: no method matching rem(::Type{BitIntegers.Int256}, ::Type{BitIntegers.UInt256})
```

Lemme add a comment to explain that.

NHDaly replied:
Done. Have another look there?

NHDaly committed:
Adjust to run for all integer types!

NHDaly committed:
Clarify the `_unsigned(x)` methods with comments

NHDaly committed:
Apply suggestions from code review

Co-authored-by: Curtis Vogt <curtis.vogt@gmail.com>

NHDaly commented:
Update 1: It turned out that (of course) `*` is too fast for `@btime` to get a good measurement of on its own. I've adjusted the benchmark to perform repeated multiplies, and the true performance is now included in the original comment.

Update 2: It turns out that this is actually faster for _all_ int types (>2x faster for FD{Int64}), even though LLVM was doing a version of this optimization on its own. I'm honestly pretty surprised by that... From what I can tell, for FD{Int32}, the code is _identical_ except for a single instruction changes from comparing `hi` to comparing `gt`:
<img width="1503" alt="Screenshot 2024-06-13 at 2 10 20 PM" src="https://github.com/JuliaMath/FixedPointDecimals.jl/assets/1582097/df9715e0-baff-4ca7-8217-33b5a11c72e4">
And yet somehow it's 25% faster (80µs -> 60µs).

NHDaly commented:
Aha, the native code difference comes from the whole benchmark function. Before, the multiply was outlined, and called through a function call.
Now, it's inlined, and i think the whole loop is constant-folded away, so it's actually sort of cheating. 😅 
But still, allowing that optimization seems like a positive thing!

omus commented on src/FixedPointDecimals.jl:
```diff
@@ -129,8 +142,10 @@ _widemul(x::Unsigned,y::Signed) = signed(_widen(x)) * _widen(y)
-_widen(::Type{UInt128}) = BitIntegers.UInt256
+_widen(::Type{Int128}) = Int256
+_widen(::Type{UInt128}) = UInt256
+_widen(::Type{Int256}) = BitIntegers.Int512
+_widen(::Type{UInt256}) = BitIntegers.UInt512
```
```suggestion
_widen(::Type{Int256}) = Int512
_widen(::Type{UInt256}) = UInt512
```

omus reviewed (commented):
Did another surface level review. I can do a full review if no one else will. I will need to block off some time though as there's a bit going on in this PR.

NHDaly committed:
Update src/FixedPointDecimals.jl

Co-authored-by: Curtis Vogt <curtis.vogt@gmail.com>

NHDaly commented:
I _think_ Tomas should be able to review when he's back from vacation. He's out til thursday. :)

NHDaly committed:
Add extensive tests for multiplication correctness, to cover the new change

Tests all FD{(U)Int16} values.
Tests most corner cases for FD{(U)Int128} values.

NHDaly committed:
Named testsets to make failures easier to identify

NHDaly committed:
Fix off-by-one error in rounding truncation in calculate_inverse_coeff()

We must use the precise value of 2^nbits(T) in order to get the correct
division in all cases.

....... UGH except now the Int64 tests aren't passing.

NHDaly committed:
Add some comments and requires

NHDaly committed:
Copy/pasted definition for unsigned numbers straight from the book

NHDaly committed:
Have `magicgu` support arbitrary integer sizes

Drvi committed:
Use the formulas from Hacker's delight for both signed and unsigned Ints

Drvi committed:
.

Drvi committed:
.

NHDaly committed:
Restrict back to just Int128 & Int256 for custom div_by_const

LLVM _does_ do this automatically if you directly call `fldmod`.

Improve the comments as well.

NHDaly committed:
It turns out that in newer versions of julia, you should call fldmod, not fld(x,y),mod(x,y), even if y is a constant! :)

Improved that for julia versions 1.8 - 1.9

NHDaly committed:
Reorganize the functions to be top-down

NHDaly committed:
More thorough tests for flmdod_by_const

NHDaly commented:
Okay, after reviewing and integrating all of Tomas' changes, and cleaning things up, i think this is a great improvement and is ready to go! :) Tomas and I are going to look through it one more time, but otherwise I think this is fully ready for review. Thanks!

NHDaly commented:
I also updated the PR comment with the final perf numbers, which look 👌

NHDaly committed:
Merge branch 'master' into nhd-Int128-fastmul-noallocs

NHDaly committed:
Bump patch version number

NHDaly committed:
Add _widemul unit test

NHDaly committed:
Change "unreachable" comment to an `@assert false`

NHDaly committed:
add test comment

EdsterG commented:
By the way, `div`/`mod` allocations are fixed in the most recent version of BitIntegers when running on 1.11+, see [#48](https://github.com/rfourquet/BitIntegers.jl/pull/48).

NHDaly commented:
Oh wow, very nice! Thanks @EdsterG. 💪 

This should still help quite a bit, but the baseline for comparison would be better in 1.11, then! 💪

NHDaly commented on src/fldmod-by-const.jl:
```diff
@@ -0,0 +1,131 @@
+    out = _widemul(promote(x, magic_number)...)
+    out >>= shift
+    # Adding one as implied by formula (1b) in Hacker's delight, Chapter 10-4.
+    return (out % T) + (x < zero(T))
```
```suggestion
    # Add one if x was negative as implied by formula (1b) in Hacker's delight, 
    # Chapter 10-4. (Note that of course if x is Unsigned, this compiles away.)
    return (out % T) + (x < zero(T))
```

NHDaly committed:
Apply suggestion from @NHDaly

Drvi committed:
Merge remote-tracking branch 'origin/master' into nhd-Int128-fastmul-noallocs

Drvi committed:
Fixes after the merge

Drvi committed:
Add JET integration tests and update fldmod_by_const tests

- Added comments
- Introduced new tests for FixedDecimal multiplication using JET.
- Enhanced fldmod_by_const tests to cover a wider range of divisors and edge cases.
- Updated Project.toml to include JET as a dependency.

Drvi committed:
Target Julia 1.10+

Drvi committed:
Improve JET tests

Drvi commented:
Ok, so I've reviewed the code and I concluded that it is correct. 

I was mainly concerned about whether should or should not add 1 for negative `x` in the last line of `div_by_const`. I later realized that, as Nathan pointed out, the +1 is very much intended and explicitly relied on in the associated proofs in Hacker's Delight. The high level reason for it is that shifts implement flooring division while we were interested in division towards zero, so for negative values, if the division is not already exact, we need to add one to compensate for the flooring. I've spent some time proving that we should always add 1 in this code path (i.e. that we don't divide evenly and that we need to compensate). I've added the proof to the body of the function as a comment, but now I think that it is not useful, since it proves a subset of what the relevant section in HD proves already (i.e. HD proves the multiplication and shift corresponds to `ceil` for negative numbers if we add 1, and `floor` for positive numbers, while I prove that given the multitplication and shift, we need to add the 1 for negative numbers).

I've also added JET tests and manually examined generated typed IR for all possible multiplication methods to verify that none of those mentioned any errors or unreachable branches. 

I've also run the exhaustive tests for `Int16`, all of them passed.

NHDaly committed:
Clean up tomas' comment

NHDaly reviewed (commented):
Alright, finally ready to go! I've moved @Drvi's last comment out to below the function, to make it easier to read. But this is finally ready to merge. I've also confirmed the performance impacts again locally.

NHDaly committed:
Fix the example benchmarks to benchmark across a wide array (to also test things like simd-compatibility)

NHDaly commented on Project.toml:
```suggestion
version = "0.6.5"
```

NHDaly replied:
Bump to 0.6.5

NHDaly committed:
Update Project.toml

NHDaly committed:
JET currently broken on nightly

NHDaly committed:
Revert "JET currently broken on nightly"

Didn't actually fix it.. I guess we can just leave the nightly tests
broken, until they fix JET? :/

This reverts commit 72156f7286683a0816d282ec3d110f75fc62f749.

NHDaly committed:
Drop JET changes
